crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/named_expr.py

```diff
@@ -3,10 +3,16 @@
 if (
     # 1
     x # 2
+    # 2.5
     := # 3
+    # 3.5
     y # 4
 ):
-    pass
+    ...
+
+# should not add brackets
```

Comment:
Hey. Just wanted to let you know that I'm working on changing `NeedsParentheses` to pass through the parent node. This should make it straightforward to implement the correct parentheses logic for the walrus operator.

Reply:
👍 

atm i can't even figure out what's causing these parens to be added.  parenthesize gets set to "ifbreaks" and nothing is breaking...

Reply:
This PR ships the infrastructure: https://github.com/astral-sh/ruff/pull/5708/

The parentheses are added because `FormatExpr` first calls into `NeedsParentheses` to determine if parentheses are necessary

https://github.com/astral-sh/ruff/blob/33a91773f7a8c60c2f1cdb84127feabad8c6d84f/crates/ruff_python_formatter/src/expression/mod.rs#L65

And the implementation of `NamedExpr` returns `Always`

https://github.com/astral-sh/ruff/blob/33a91773f7a8c60c2f1cdb84127feabad8c6d84f/crates/ruff_python_formatter/src/expression/expr_named_expr.rs#L39-L44

We should implement the same rules as in https://github.com/psf/black/blob/d1248ca9beaf0ba526d265f4108836d89cf551b7/src/black/linegen.py#L1381-L1395 (requires access to the parent node)

Reply:
aah. thanks

do i understand correctly that his is separate to (i.e. won't help with) parens for generator comprehensions (that we were discussing yesterday)?